lib/mongo/retryable.rb

```diff
@@ -548,5 +496,51 @@ def raise_unsupported_error(e)
       new_error.set_backtrace(e.backtrace)
       raise new_error
     end
+
+    def retryable_exceptions
+      [
+        Error::SocketError,
+        Error::SocketTimeoutError,
+        Error::ServerNotUsable,
+        Error::ConnectionPerished
+      ].freeze
+    end
+
+    def is_retryable_exception?(exception)
+      retryable_exceptions.any? { |klass| klass === exception }
+    end
+
+    def retry_legacy_read_with_retry(exception, attempt, session)
+      exception.add_note('legacy retry')
+      exception.add_note("attempt #{attempt}")
+
+      if is_retryable_exception?(exception)
+        if attempt > client.max_read_retries || session&.in_transaction?
+          raise exception
+        end
+      elsif exception.retryable? && !session&.in_transaction?
+        if attempt > client.max_read_retries
+          raise exception
+        end
+      else
+        raise exception
+      end
+
+      log_retry(exception, message: 'Legacy read retry')
+      sleep(client.read_retry_interval) unless is_retryable_exception?(exception)
+    end
+
+    def prepare_to_retry_legacy_write_with_retry(exception, attempt)
```

Comment:
According to its usage - https://github.com/mongodb/mongo-ruby-driver/pull/2682/files?diff=split&w=1#diff-d01924119d9e1f082918101e83057a7fac7ff818e8126320fec1714d308b73f0R350 - this method should return `server`. However, if I am not wrong, it either return `nil`, or raise.
I believe that the idea is that returning `nil` forces a server selection; this seems to be a bit obscure.

Reply:
Yeah, I could document this better. I considered making it return nothing, and explicitly setting `server` to `nil` at the callsite, but I figured I could kill two birds with one stone here. :) I was probably trying to be too clever. I'll clarify.